Kernel greedy option and Theory documentation (#3)
efekhari27/otkerneldesign: doc/theory/theory.rst

functional space were used to approximate integrals. More precisely, one can mention the use 
of the distance called the *maximum mean discrepancy* (MMD) as a core ingredient of advanced sampling 
methods such as the *Support points* by (Mak & Joseph, 2018) and the *Kernel herding* by (Chen & al., 2010). 
Manipulating the MMD is convenient since its expression is simple ; it depends on an arbitrarily chosen kernel.  
Let us setup the introduction of the Kernel herding and Support points methods by briefly defining a few mathematical concepts. 

**Reproducing kernel Hilbert space.**

Assuming that :math:`k` is a symmetric and positive definite function :math:`k: \cD_\vect{X} \times \cD_\vect{X} \rightarrow \Rset`, 
latter called a "reproducing kernel" or simply a "kernel". A *reproducing kernel Hilbert space* (RKHS) is an inner product 
space :math:`\cH(k)` of functions :math:`g:\cD_\vect{X} \rightarrow \Rset` with the following properties:

* :math:`k(\cdot, \vect{x}) \in \cH(k), \quad \forall \vect{x} \in \cD_\vect{X}.`
* :math:`\langle g, k(\cdot, \vect{x}) \rangle_{\cH(k)} = g(\vect{x}), \quad \forall \vect{x} \in \cD_\vect{X}, \forall g \in \cH(k)`

> Notice that for a defined reproducing kernel, a unique RKHS exists and vice versa (see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ).

Reviewer: ```suggestion
Note that for a defined reproducing kernel, a unique RKHS exists and vice versa (see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ).
```
The original formulation implies that this is trivial.

**Potential.**

For any target distribution :math:`\mu`, its *potential* (also called "kernel mean embedding") associated with the kernel :math:`k` is defined as: 

.. math::
    :name: potential
    
    P_{\mu}(\vect{x}) := \int_{\cD_\vect{X}} k(\vect{x}, \vect{x}') \di \mu(\vect{x}').

Then, the potential of a discrete distribution :math:`\zeta_n = \frac1n \sum_{i=1}^{n} \delta(\vect{x}^{(i)})` 
(uniform mixture of Dirac distributions at the design points :math:`\vect{X}_n`) associated with the kernel :math:`k` can be expressed as:

.. math::
    :name: design_potential
    
    P_{\zeta_n}(\vect{x}) = \frac1n \sum_{i=1}^{n} k(\vect{x}, \vect{x}^{(i)}).

> Close potentials can be interpreted to mean that the design $\bX_n$ adequately quantizes $\mu$

Reviewer: ```suggestion
Close potentials can be interpreted to mean that the design :math:`\bX_n` adequately quantizes :math:`\mu`.
```

**Maximum mean discrepancy.**

> A recent metric of discrepancy and quadrature error is offered by the MMD. 

Reviewer: Is it recent?

This distance between two distributions :math:`\mu` and :math:`\zeta` is given by the 

> maximal quadrature error committed for any function within a defined RKHS:

Reviewer: ```suggestion
maximal quadrature error committed for any function within the unit ball of an RKHS:
```

.. math::
    :name: mmd
    
    \mathrm{MMD}_k(\mu, \zeta) := 
    \sup_{\lVert g \lVert_{\cH(k)} \leq 1}
            \left | \int_{\cD_{\vect{X}}} g(\vect{x}) \di \mu(\vect{x}) - \int_{\cD_{\vect{X}}} g(\vect{x}) \di \zeta(\vect{x}) \right|.

Using the Cauchy-Schwartz inequality, one can demonstrate that 
:math:`\mathrm{MMD}_k(\mu, \zeta) = \left\lVert P_{\mu}(\vect{x}) - P_{\zeta}(\vect{x}) \right\lVert_{\cH(k)}` 
(see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ). 

> Moreover, kernels are called characteristic when the following implication is true , :math:`\mathrm{MMD}_k(\mu, \zeta) = 0 \Rightarrow \mu = \zeta`.

Reviewer: ```suggestion
Moreover, a kernel :math:`k` is called characteristic if :math:`\mathrm{MMD}_k(\mu, \zeta) = 0` is equivalent to :math:`\mu = \zeta`.
```

(see (Pronzato & Zhigljavsky, 2020) for the full proof). 

In practice, :math:`P_{\mu}(\vect{x})` can be expressed analytically in the specific cases of input distribution and kernel 
(e.g., for independent uniform or normal inputs and a Matérn :math:`5/2` kernel (Fekhari & al., 2021)), making the computation very fast. 
Alternatively, the potential can be evaluated on an empirical measure :math:`\mu_N`, substituting :math:`\mu`, 
formed by a dense and large-size sample of :math:`\mu` (e.g., the candidate set :math:`\mathcal{S}`). 
:math:`P_{\mu}(\vect{x})` is then approached by :math:`P_{\mu_N}(\vect{x}) = (1/N)\, \sum_{j=1}^N k(\vect{x}, \vect{x}'^{(j)})`, 
which can be injected in :ref:`(4) <kh_criterion>` to solve the following optimization:

.. math::
  :name: kh_estimation

    \vect{x}^{(n+1)} \in \argmin_{\vect{x}\in\mathcal{S}} \left( \frac{1}{n} \sum_{i=1}^{n} k(\vect{x},\vect{x}^{(i)}) 
    - \frac{1}{N} \sum_{j=1}^N k(\vect{x},\vect{x}'^{(j)}) \right) \,.

When no observation is available, which is the common situation at the design stage, 
the kernel hyperparameters (e.g., correlation lengths) have to be set to heuristic values. 

> MMD minimization is quite versatile and was explored in more details by (Teymur & al., 2021)

Reviewer: ```suggestion
MMD minimization is quite versatile and was explored in detail by (Teymur & al., 2021)
```
Not sure about this one. What do you think?

Compared to more heuristic methods for solving quantization problems, Support points
benefit from the theoretical guarantees of MMD minimization in terms of convergence of :math:`\zeta_n` to :math:`\mu` as :math:`n\to\infty`. 

At first sight, this optimization problem seems intractable, although (Mak & Joseph, 2018) propose to 
rewrite the function as a difference of convex functions in :math:`\vect{X}_n`, which yields a difference-of-convex program. 
To simplify the algorithm and keep an iterative design, a different approach will be used here. 
At iteration :math:`n+1`, the algorithm solves greedily the MMD minimization between :math:`\zeta_n` and :math:`\mu` for the candidate set :math:`\mathcal{S}`:

.. math::
    :name: greedy_criterion

    \vect{x}^{(n+1)} \in \argmin_{\vect{x}\in\mathcal{S}} \Bigg( \frac{1}{N} \sum_{j=1}^N \|\vect{x}-\vect{x}'^{(j)}\| 
    - \frac{1}{n+1} \sum_{i=1}^{n} \|\vect{x}-\vect{x}^{(i)}\| \Bigg) \,.

For this criterion, one can notice that it is almost identical to the KH one in :ref:`(4) <kh_criterion>` when 
taking as kernel the energy-distance kernel given in :ref:`(6) <energy_kernel>`.

> These two iterative methods were exploited in (Fekhari & al., 2021) to study new ways to construct 

Reviewer: Maybe add link to PDFs of articles when possible?
